app/codes/validator.py

```diff
@@ -45,3 +50,70 @@ def validate(transaction):
     status = f"Wrote check status as {check} to {checkfile}"
     print(status)
     return status
+
+
+def validate_signature(data, public_key, signature):
+    public_key_bytes = base64.b64decode(public_key)
+    sign_trans_bytes = base64.decodebytes(signature.encode('utf-8'))
+    vk = ecdsa.VerifyingKey.from_string(
+        public_key_bytes, curve=ecdsa.SECP256k1)
+    message = json.dumps(data).encode()
+    try:
+        return vk.verify(sign_trans_bytes, message)
+    except:
+        return False
+
+
+def validate_receipt_signature(receipt):
+    try:
+        return validate_signature(receipt['data'], receipt['public_key'], receipt['signature'])
+    except:
+        logger.error('Error validating receipt signature')
+        return False
+
+
+def get_node_trust_score(public_key):
+    # TODO - Return the actual trust score of the node by lookup on public_key
+    return 1
+
+def validate_block_using_receipts(block):
+    total_receipt_count = 0
+    score_weighted_validity_count = 0
+
+    for receipt in block['receipts']:
+        total_receipt_count += 1
+
+        if not validate_receipt_signature(receipt):
+            continue
+
+        if receipt['data']['block_index'] != block['index'] or receipt['data']['block_hash'] != block['hash'] or receipt['data']['vote'] < 1:
+            continue
+
+        trust_score = get_node_trust_score(receipt['public_key'])
+        valid_probability = 0 if trust_score < 0 else (trust_score + 2) / 5
```

Comment:
@swapnilpawar31415 critical part here

Reply:
Will re-write the logic in the white-paper for this part and send to you.
Need to separate the information contained in a negative vote vs no vote, which is not achieved currently.